server/src/operators/chunk_operator.rs

```diff
@@ -1548,21 +1555,60 @@ pub fn get_highlights_with_exact_match(
         })
         .collect_vec();
 
+    let mut exact_match_idx_set = HashSet::new();
+    for (idx, _) in &exact_matches {
+        exact_match_idx_set.insert(idx);
+    }
+
+    dbg!(&exact_matches);
+
+    dbg!(&matched_idxs_with_windows);
+
     let merged_results = matched_idxs_with_windows
         .chunk_by(|(_, _, _, _, a_w_end), (_, _, _, b_w_start, _)| a_w_end >= b_w_start)
-        .filter_map(|merged_run| match (merged_run.first(), merged_run.last()) {
-            (Some((_, _, _, first_start, _)), Some((_, _, _, _, last_end))) => {
-                let combined_phrases = merged_run
-                    .iter()
-                    .map(|(_, _, phrase, _, _)| phrase.to_string())
-                    .unique()
-                    .collect_vec();
-                Some((first_start, last_end, combined_phrases))
+        .flat_map(|merged_run| {
+            let mut result = Vec::new();
+            let mut prev_was_exact = exact_match_idx_set.contains(&merged_run[0].0);
+            let mut prev_w_start = merged_run[0].3;
+            let mut prev_w_end = merged_run[0].4;
+            let mut combined_phrases = vec![merged_run[0].2.to_string()];
+            for run in merged_run.iter().skip(1) {
+                let is_exact = exact_match_idx_set.contains(&run.0);
+                if is_exact == prev_was_exact && is_exact {
+                    // prev_w_end = run.w_end;
+                    prev_w_end = run.4;
+                    combined_phrases.push(run.2.to_string());
+                } else {
+                    if prev_was_exact {
+                        result.push((
+                            prev_w_start,
+                            prev_w_end,
+                            combined_phrases.clone().into_iter().unique().collect_vec(),
+                        ));
+                    } else {
+                        result.push((
+                            prev_w_start,
+                            run.1.saturating_sub(1).min(prev_w_end),
+                            combined_phrases.clone().into_iter().unique().collect_vec(),
+                        ))
+                    }
+                    prev_w_start = run.3.max(prev_w_end).min(run.1);
+                    prev_w_end = run.4;
+                    combined_phrases = vec![run.2.to_string()];
+                }
+                prev_was_exact = is_exact;
```

Comment:
trying to truncate windows selctively here, need to correctly decide what `prev_w_start` should be and what the `else` branch should be pushing to get the results we want